pkg/column-mapping/column.go

```diff
@@ -0,0 +1,321 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package column
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb-tools/pkg/table-rule-selector"
+)
+
+// Expr indicates how to handle column mapping
+type Expr string
+
+// poor Expr
+const (
+	AddPrefix Expr = "add prefix"
+	AddSuffix Expr = "add suffix"
+	Clone     Expr = "clone column"
+)
+
+// Exprs is some built-in expression for column mapping
+// only support some poor expressions now, we would unify tableInfo later and support more
+var Exprs = map[Expr]func(*columnInfo, []interface{}) []interface{}{
+	AddPrefix: addPrefix,
+	AddSuffix: addSuffix,
+	Clone:     cloneColumn,
+}
+
+// Rule is a rule to map column
+type Rule struct {
+	PatternSchema    string   `yaml:"pattern-schema" json:"pattern-schema" toml:"pattern-schema"`
+	PatternTable     string   `yaml:"pattern-table" json:"pattern-table" toml:"pattern-table"`
+	SourceColumn     string   `yaml:"source-column" json:"source-column" toml:"source-column"` // modify, add refer column, ignore
+	TargetColumn     string   `yaml:"target-column" json:"target-column" toml:"target-column"` // add column, modify
+	Expression       Expr     `yaml:"expression" json:"expression" toml:"expression"`
+	Arguments        []string `yaml:"arguments" json:"arguments" toml:"arguments"`
+	CreateTableQuery string   `yaml:"create-table-query" json:"create-table-query" toml:"create-table-query"`
+}
+
+// Valid checks validity of rule.
+// add prefix/suffix: it should have target column and one argument
+// clone: it should have source and target column
+func (r *Rule) Valid() error {
+	if _, ok := Exprs[r.Expression]; !ok {
+		return errors.NotFoundf("expression %s", r.Expression)
+	}
+
+	if (r.Expression == AddPrefix || r.Expression == AddSuffix) && len(r.Arguments) != 1 {
+		return errors.NotValidf("arguments %v for add prefix/suffix", r.Arguments)
+	}
+
+	return nil
+}
+
+// check source and target position
+func (r *Rule) checkColumnPosition(source, target int) (int, int, error) {
+	// if not found target, ignore it
+	if target == -1 {
+		return source, target, errors.NotFoundf("target column %s", r.TargetColumn)
+	}
+
+	if r.Expression == Clone && source == -1 {
+		return source, target, errors.NotFoundf("source column %s", r.SourceColumn)
+	}
+
+	if target < source { // must add column and added column is before source column
+		source--
```

Comment:
should we need to test `Expression`? if we added more expression some day, it is possible not a **add** operation, but is still `target < source`

Reply:
now expression is too simple,  I would put it within L75 block